demos/gplates_global/forward.py

```diff
@@ -0,0 +1,370 @@
+    r = sqrt(X[0]**2 + X[1]**2 + X[2]**2)
+
+    # Depth dependence: for the lower mantle increase we
+    # multiply the profile with a linear function
+    for line, step in zip([5.*(rmax-r), 1., 1.],
+                          [step_func(r, 1.992, 30, False),
+                           step_func(r, 2.078, 10, False),
+                           step_func(r, 2.2, 10, True)]):
+        mu_lin += line*step
+
+    # Adding temperature dependence:
+    delta_mu_T = Constant(100.)
+    mu_lin *= exp(-ln(delta_mu_T) * T)
+    mu_star, sigma_y = Constant(1.0), 5.0e5 + 2.5e6*(rmax-r)
+    epsilon = sym(grad(u))  # strain-rate
+    epsii = sqrt(inner(epsilon, epsilon) + 1e-10)  # 2nd invariant (with a tolerance to ensure stability)
+    mu_plast = mu_star + (sigma_y / epsii)
+    mu = (2. * mu_lin * mu_plast) / (mu_lin + mu_plast)
+    return mu
+
+
+def assign_1d_profile(q, one_d_filename):
+    """
+    Assign a one-dimensional profile to a Function `q` from a file.
+
+    The function reads a one-dimensional radial viscosity profile from a file, broadcasts
+    the read array to all processes, and then interpolates this
+    array onto the function space of `q`.
+
+    Args:
+        q (firedrake.Function): The function onto which the 1D profile will be assigned.
+        one_d_filename (str): The path to the file containing the 1D radial viscosity profile.
+
+    Returns:
+        None: This function does not return a value. It directly modifies the input function `q`.
+
+    Note:
+        - This function is designed to be run in parallel with MPI.
+        - The input file should contain an array of viscosity values.
+        - It assumes that the function space of `q` is defined on a radial mesh.
+        - `rmax` and `rmin` should be defined before this function is called, representing
+          the maximum and minimum radial bounds for the profile.
+    """
+    import numpy as np
+    from firedrake.ufl_expr import extract_unique_domain
+    from scipy.interpolate import interp1d
+    # find the mesh
+    mesh = extract_unique_domain(q)
+
+    visc = None
+    rshl = None
+    # read the input file
+    if mesh.comm.rank == 0:
+        # The root process reads the file
+        rshl, visc = np.loadtxt(one_d_filename, unpack=True, delimiter=",")
+
+    # Broadcast the entire 'visc' array to all processes
+    visc = mesh.comm.bcast(visc, root=0)
+    # Similarly, broadcast 'rshl' if needed (assuming all processes need it)
+    rshl = mesh.comm.bcast(rshl, root=0)
+
+    element_family = q.function_space().ufl_element()
+    X = Function(VectorFunctionSpace(mesh=mesh, family=element_family)).interpolate(SpatialCoordinate(mesh))
+    rad = Function(q.function_space()).interpolate(sqrt(X**2))
+    averager = LayerAveraging(mesh, cartesian=False)
+    averager.extrapolate_layer_average(q, interp1d(rshl, visc, fill_value="extrapolate")(averager.get_layer_average(rad)))
+
+
+if __name__ == "__main__":
+    forward()
```

Comment:
Please run this example, get a set of diagnostic values at the end of the simulation and then add it to the testing environment to ensure ongoing functionality.

Reply:
Didn't we want to move this level of testings from now on to testing_long directory? @angus-g what do you say?

Reply:
I can't actually run this as-is:
```
Traceback (most recent call last):
  File "/home/firedrake/g-adopt/demos/gplates_global/forward.py", line 370, in <module>
    forward()
  File "/home/firedrake/g-adopt/demos/gplates_global/forward.py", line 77, in forward
    muller_2022_files = obtain_Muller_2022_SE("./gplates_files")
  File "/home/firedrake/g-adopt/gadopt/gplatefiles.py", line 37, in obtain_Muller_2022_SE
    raise FileNotFoundError("Some files are missing and download_mode is set to False. Cannot proceed without downloading the required files.")
FileNotFoundError: Some files are missing and download_mode is set to False. Cannot proceed without downloading the required files.
```

Having said that, I think this is big enough that we only want it in the long test suite? If you give me an idea of core count and runtime, I can wire that up.

Reply:
Yeah! So part of the run case should be calling `obtain_Muller_2022_SE("./gplates_files", download_mode=True)` from `gadopt.gplatefiles`. And then calling `forward()`, noting that we call the `obtain_Muller_2022_SE("./gplates_files", download_mode=False)` on  https://github.com/g-adopt/g-adopt/blob/26ad2bd8c2ed47bd2106ba2096548207de79dd51/demos/gplates_global/forward.py#L77

This is basically a level 6 spherical case. So only in long suite. Core count should be the same as level 6 in other spherical case, with the same runtime (here we run it for only 20 timesteps). I will give more accurate numbers later.

Reply:
Thanks! That's enough for me to go on for now :)